### PR TITLE
Change sticky header position fixed to absolute and set its top position relevant to vertical scroll bar

### PR DESCRIPTION
**Pull request by sachitp**
This PR is raised for issue:

sticky-table-header mis aligns when the parent container is given some fixed width and not set to 100%. #39

### CODE REVIEW

**Comment by hannaxbear** [comment]
Hi @bcherny, this PR fixes the bug that we ran into with Datahub, where the sticky header would overflow outside of the fixed width container. By changing this to position: absolute, the header will respect the position and width of the parent container. I tested this fix with Datahub and the other tables and it works well.

**Comment by bcherny** [comment]
@hannaxbear looks good. there's no perf hit or lag from repositioning the y on scroll?

**Comment by hannaxbear** [comment]
@bcherny , hmm good point. It does look like the header flickers a bit when scrolling quickly..

**Comment by bcherny** [comment]
@hannaxbear instead, can you do something like this in the consumer? http://stackoverflow.com/a/5219023/435124

**Comment by hannaxbear** [comment]
@sachitp , can you look into a solution that may provide better performance when scrolling?

**Comment by hannaxbear** [comment]
@bcherny , after taking a second look at the performance, it seems that the flickering is difficult to reproduce and rather subtle even when it is there. I think that it should not be a concern. What do you think?

Also, @sachitp looked at the stackoverflow post you linked and said: "It seems like problem posted is different from what we have, Elements(div) mentioned in posted problem are siblings whereas we have parent child relation so this solution does not applies in our case"

**Comment by bcherny** [comment]
@hannaxbear can you guys update your dom so the fix from the so post works? the sticky header already has pretty poor performance, i wouldn't want to make it even worse.

**Comment by hannaxbear** [comment]
@bcherny , I have finally gotten the chance to investigate this issue further.

The problem is that we have a parent div that has a given width with overflow: scroll, where we expect the content inside this div to be hidden from view with a scrollbar if it exceeds the given width. The sticky header lives inside this parent div, but since it has position:fixed, it does not obey the parent div's width/overflow properties (this SO answer explains this nicely: http://stackoverflow.com/a/12463736). 

So, to answer your question, the SO post you referenced would not solve this problem. That post is dealing with 2 divs: (1) a div with position: fixed, and (2) a div with position: static that we want to display on top of div 1. But in our situation, we aren't trying to display a position: static div on top of a position: fixed div. We just want to the position: fixed div to hide its overflow. Do you see what I mean?

Now, going back to the other SO post I linked above, there is another answer that provides a possible solution: http://stackoverflow.com/a/23859719. However, along with some other caveats, it requires that the parent div be position: absolute, which is impractical for most use cases, including ours, so I don't think it would make sense to use this solution.

Aside from that, I think the solution @sachitp presented in this PR is the only working solution. He also investigated the UX experience further with the following findings:
> 1. This flickering is noticeable in chrome only while scrolling the page using middle mouse wheel.
2. In Chrome It seems like the re positioning of the div happens somewhat delayed, as if the page is scrolled first and then the div is moved to the new position.
3. Providing few links:
http://www.tagwith.com/question_791327_chrome-flickers-when-scrolling-with-mouse-wheel-relocating-div-to-document-s/
https://code.google.com/p/chromium/issues/detail?id=419279

In conclusion, I think this is our only option, and if you'd like, perhaps we can introduce an optional setting that will use position: fixed by default and the developer using angular-sticky-header can set it to use position: absolute for these types of situations. What do you think?

File: dist/angular-sticky-table-header.js [hunk 1]
```diff
@@ -109,7 +109,8 @@ angular.module('turn/stickyTableHeader', ['watchDom']).value('stickyTableHeaderO
             } else if (scope.stuck && scrollY < scope.offset.top) {
               scope.setStuck(false);
             }
-            scope.clone.css('left', scope.offset.left - scrollX);
+            scope.clone.css('left', 0);
+            scope.clone.css('top', ($window.scrollY || 0) - scope.offset.top);
```
Comment (bcherny) [inline dist/angular-sticky-table-header.js]: please use `pageYOffset` instead of `scrollY` for ie compatability

File: dist/angular-sticky-table-header.js [hunk 2]
```diff
@@ -109,7 +109,8 @@ angular.module('turn/stickyTableHeader', ['watchDom']).value('stickyTableHeaderO
             } else if (scope.stuck && scrollY < scope.offset.top) {
               scope.setStuck(false);
             }
-            scope.clone.css('left', scope.offset.left - scrollX);
```
Comment (bcherny) [inline dist/angular-sticky-table-header.js]: this breaks horizontal scrolling, why did you remove this? or does abs positioning make it ok?

**Comment by bcherny** [comment]
@hannaxbear i made a couple of code comments, otherwise lgtm. please add unit tests too.

File: dist/angular-sticky-table-header.js (hunk 1 above)
Comment (jwalin) [inline dist/angular-sticky-table-header.js]: Fixed. Used pageYOffset instead of scrollY.

Comment (sachitp) [inline dist/angular-sticky-table-header.js]: Fixed. Used pageYOffset instead of scrollY.

File: dist/angular-sticky-table-header.js (hunk 2 above)
Comment (sachitp) [inline dist/angular-sticky-table-header.js]: @bcherny .  Absolute positioning handle horizontal scrolling. I checked it is not breaking.

**Comment by sachitp** [comment]
Added unit test.

File: dist/angular-sticky-table-header.js (hunk 2 above)
Comment (bcherny) [inline dist/angular-sticky-table-header.js]: in that case, can you just remove line 112?

Comment (sachitp) [inline dist/angular-sticky-table-header.js]: @bcherny . The horizontal scroll is taken care by setting position to absolute ALONG with left to 0. Both position: absolute & left: 0 has to go hand in hand in order to have the horizontal scroll work seamlessly.

Comment (bcherny) [inline dist/angular-sticky-table-header.js]: can we just set `left:0` in the css instead of having to re-set it on every tick?

Comment (sachitp) [inline dist/angular-sticky-table-header.js]: @bcherny If we set left:0 from css directly then while scrolling left position is override by setCloneGutter() which is called prior to checkScroll() and  header left position is set to some value instead of 0.

Hence, we explicitly need to set left position to 0 in checkScroll().

Comment (bcherny) [inline dist/angular-sticky-table-header.js]: @sachitp then the answer is to not set it in `setCloneGutter`, rather than set it and then override it, right?...

Comment (sachitp) [inline dist/angular-sticky-table-header.js]: @bcherny . Ok. We have removed code which update the left position in `setCloneGutter`.
Set `left:0` in css instead of having to re-set it on every tick.

File: test/test.coffee [hunk 3]
```diff
@@ -436,7 +435,7 @@ describe 'angular-sticky-table-header', ->
 			.not.toHaveBeenCalled
 
 
-		it 'should set left', ->
+		it 'should not set left', ->
```
Comment (bcherny) [inline test/test.coffee]: This test isn't useful anymore, just delete it.

Comment (sachitp) [inline test/test.coffee]: Deleted unnecessary test.

**Comment by bcherny** [comment]
Lgtm

**Comment by bcherny** [comment]
I'll merge this guy when I have a computer and can update the guy tags and republish this, next week. Sorry about the delay.

**Comment by hannaxbear** [comment]
@bcherny , any updates on this? Thanks.

**Comment by bcherny** [comment]
@hannaxbear merged

**Comment by hannaxbear** [comment]
Thanks @bcherny !
